website/static/js/fangorn.js

```diff
@@ -1013,6 +996,491 @@ function scrollToFile(fileID) {
+
+var FGInput = {
+    controller : function(args) {
+        var noop = function() {};
+    },
+    view : function(ctrl, args, helpText) {
+        var extraCSS = args.className || '';
+        var tooltipText = args.tooltip || '';
+        var placeholder = args.placeholder || '';
+        var id = args.id || '';
+        var helpTextId = args.helpTextId || '';
+        return m('span', [
+            m('input', {
+                'id' : id,
+                className: 'tb-header-input' + extraCSS,
+                onclick: args.onclick,
+                onkeydown: args.onkeydown,
+                'data-toggle':  tooltipText ? 'tooltip' : '',
+                'title':  tooltipText,
+                'data-placement' : 'bottom',
+                'placeholder' : placeholder
+                }),
+            m('.text-danger', {
+                'id' : helpTextId
+            }, helpText)
+        ]);
+    }
+}
+
+var FGDropdown = {
+    controller : function(args) {
+        var noop = function() {};
+    },
+    view : function(ctrl, args, children) {
+        var extraCSS = args.className || '';
+        var tooltipText = args.tooltip || '';
+        var id = args.id || '';
+        var name = args.name || '';
+        var label = args.label || '';
+        return m('div', {
+                className: 'fangorn-toolbar-icon ' + extraCSS
+            },[
+                m('span.hidden-xs',label),
+                m('select.no-border', {
+                    'name' : name,
+                    'id' : id,
+                    onchange: args.onchange,
+                    'data-toggle':  tooltipText ? 'tooltip' : '',
+                    'title':  tooltipText,
+                    'data-placement' : 'bottom'
+                },children)
+        ]);
+    }
+}
+
+var _dismissToolbar = function(){
+    var tb = this;
+    tb.toolbarMode('bar');
+    tb.resetFilter();
+    tb.filterText('');
+};
+
+var FGToolbar = {
+    controller : function(args) {
+        var self = this;
+        self.tb = args.treebeard;
+        self.tb.toolbarMode = m.prop('bar');
+        self.items = args.treebeard.multiselected;
+        self.mode = self.tb.toolbarMode;
+        self.uploadState = args.treebeard.isUploading;
+        self.helpText = m.prop('');
+        self.dismissToolbar = _dismissToolbar.bind(self.tb);
+        self.createFolder = function(event){
+            _createFolder.call(self.tb, event, self.dismissToolbar, self.helpText );
+        };
+    },
+    view : function(ctrl) {
+        var templates = {};
+        var generalButtons = [];
+        var rowButtons = [];
+        var items = ctrl.items();
+        var item = items[0];
+        var dismissIcon = m.component(FGButton, {
+                onclick: ctrl.dismissToolbar,
+                tooltip: 'Close Search',
+                icon : 'fa fa-times'
+            }, '');
+        templates.search =  [
+            m('.col-xs-10', [
+                ctrl.tb.options.filterTemplate.call(ctrl.tb)
```

Comment:
In the future, we should discuss further refactoring of Treebeard to convert these template functions into proper components.